// bluesdk-url/build/url.js
'use strict'

Object.defineProperty(exports, '__esModule', {
  value: true
})
exports.default = {

  // 获取域名的根域名
  getRootDomain: function getRootDomain (domain) {
    domain = domain || window.location.host
    var match = domain.match(/\w+\.\w+$/)
    if (match) {
      return match[0]
    }
    match = domain.match(/\w+\.\w+:\d+$/)
    if (match) {
      return match[0]
    }
    return window.location.host
  },

  getQueryStr: function getQueryStr (url) {
    if (!url) url = location.href
    var match = url.match(/\?(.+)/)
    return match ? match[1] : ''
  },

  getQuery: function getQuery (url) {
    return this.strToQuery(this.getQueryStr(url))
  },

  getPara: function getPara (key, defaultValue) {
    var query = this.getQuery()
    return query && query[key] ? query[key] : defaultValue === undefined ? '' : defaultValue
  },

  replaceQuery: function replaceQuery (newQuery) {
    var query = this.getQuery()
    for (var i in newQuery) {
      if (!newQuery.hasOwnProperty(i)) continue
      query[i] = newQuery[i]
    }
    return query
  },

  strToQuery: function strToQuery (queryStr) {
    var query = {}
    if (!queryStr) return query
    var items = queryStr.split('&')
    for (var index in items) {
      if (!items.hasOwnProperty(index)) continue
      var item = items[index].split('=')
      if (item.length === 2) {
        query[item[0]] = decodeURIComponent(item[1]).trim()
      }
    }
    return query
  },
  queryToStr: function queryToStr (query) {
    var r = []
    for (var i in query) {
      if (!query.hasOwnProperty(i) || query[i] === undefined) continue
      r.push(i + '=' + encodeURIComponent(query[i].toString().trim()))
    }
    return r.join('&')
  }

}
