// bluesdk-mask-layer/src/maskLayer.js
import $ from 'jquery';
export default {
    show: function (duration) {
        if ($('#_maskLayer').length === 0) {
            $('body').prepend('<div id="_mask_layer" class="_bluesdk_notifier"></div>')
                .find('#_mask_layer')
                .css({
                    opacity: 0.5,
                    background: 'black',
                    position: 'absolute',
                    top: 0,
                    zIndex: 998,
                   // width: '100%',
                    width: $(document).width(),
                    height: $(document).height(),
                    display: 'none'
                })

        } else {
            $('#_mask_layer').height($(document).height());
        }
        if (duration) {
            $('#_mask_layer').fadeIn(duration);
        } else {
            $('#_mask_layer').show();
        }
    },
    hide: function (duration) {
        if (duration) {
            $('#_mask_layer').fadeOut(duration);
        } else {
            $('#_mask_layer').hide();
        }
    }

}
