// bluesdk-upload/build/AjaxUpload.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _Html5Upload = require('./Html5Upload.js');

var _Html5Upload2 = _interopRequireDefault(_Html5Upload);

var _SWFUpload = require('./SWFUpload.js');

var _SWFUpload2 = _interopRequireDefault(_SWFUpload);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var AjaxUpload = {

    QUEUE_ERROR: {
        QUEUE_LIMIT_EXCEEDED: -100,
        FILE_EXCEEDS_SIZE_LIMIT: -110,
        ZERO_BYTE_FILE: -120,
        INVALID_FILE_TYPE: -130
    },

    buttonAction: {
        SELECT_FILE: -100,
        SELECT_FILES: -110,
        START_UPLOAD: -120
    },
    UPLOAD_ERROR: {
        HTTP_ERROR: -200,
        MISSING_UPLOAD_URL: -210,
        IO_ERROR: -220,
        SECURITY_ERROR: -230,
        UPLOAD_LIMIT_EXCEEDED: -240,
        UPLOAD_FAILED: -250,
        SPECIFIED_FILE_ID_NOT_FOUND: -260,
        FILE_VALIDATION_FAILED: -270,
        FILE_CANCELLED: -280,
        UPLOAD_STOPPED: -290
    },
    CURSOR: {
        ARROW: -1,
        HAND: -2
    },
    WINDOW_MODE: {
        WINDOW: "window",
        TRANSPARENT: "transparent",
        OPAQUE: "opaque"
    },
    instanceNum: 0,
    createUploader: function createUploader(settings) {
        settings.instance_id = ++this.instanceNum;
        if (settings['mode'] === undefined) {
            settings['mode'] = window.FormData === undefined ? 'flash' : 'html5';
        }
        if (settings['mode'] === 'flash') {
            if (settings['flash_url'] === undefined) {
                settings['flash_url'] = require("./swfupload.swf");
            }
            if (settings['button_image_url'] === undefined) {
                settings['button_image_url'] = require("./swfuploadBtn.png");
            }
            return new _SWFUpload2.default(settings);
        } else {
            settings['fileSizeLimit'] = settings['file_size_limit'];
            return new _Html5Upload2.default(settings, AjaxUpload);
        }
    }
};

exports.default = AjaxUpload;
