// cpc-frontend-base/build/util/ForeignUtil.js
import CurrencyMap from '../enum/CurrencyMap';

/**
 * 跨境交易 工具方法
 */
export default {
    /**
     * 是否跨境人民币汇款
     * @param currency
     */
    isRmbRemit: function isRmbRemit(currency) {
        return currency === 'CNY';
    },
    /**
     * 显示汇单币种
     * @param payType
     * @param fixMoneyType
     * @param currency
     * @returns {string}
     */
    orderCurrency: function orderCurrency(currency, fixMoneyType) {
        return this.isRmbRemit(currency) || fixMoneyType.toString() === '0' ? '人民币' : CurrencyMap[currency];
    },
    payeePersonalIdNumberLabel: function payeePersonalIdNumberLabel(isCompany) {
        return isCompany ? '企业法人身份证号码' : '收款人身份证号码';
    },
    payeePhoneNumberLabel: function payeePhoneNumberLabel(isCompany) {
        return isCompany ? '企业法人手机号' : '收款人手机号';
    },
    payeePersonalFirstNameLabel: function payeePersonalFirstNameLabel(isCompany) {
        return isCompany ? '企业法人名字' : '收款人名字';
    },
    payeePersonalLastNameLabel: function payeePersonalLastNameLabel(isCompany) {
        return isCompany ? '企业法人姓氏' : '收款人姓氏';
    }
};
